common/src/stack/command/stack/commands/set/host/power/__init__.py

```diff
@@ -8,96 +8,139 @@
-
-		if cmd == 'status':
-			#
-			# used by "stack list host power" -- this is an easy way in which to
-			# share code between the two commands
-			#
-			# set 'debug' to True in order to get output from the status command
-			#
-			debug = True
-		elif cmd not in [ 'on', 'off', 'reset' ]:
-			raise ParamError(self, 'command', 'must be "on", "off" or "reset"')
-		
+		cmd, debug, force_imp = self.fillParams([
+			('command', None, True),
+			('debug', False),
+			('use-method', None)
+		])
+		imp_names = []
 		self.debug = self.str2bool(debug)
-		
+
+		# Flag for if an implementation has
+		# successfully run
+		imp_success = False
+
+		self.loadImplementation()
+
+		# The first implementation is ipmi by default
+		# but can be forced by the use-method parameter
+		# If this is set, only that implementation will be run
+		if force_imp and force_imp in self.impl_list:
+			imp_names.append(force_imp)
+
+		# Otherwise use all the implementations
+		else:
+			# Gather all set power implmentations besides ipmi and ssh
+			imp_names  = [imp for imp in self.impl_list if 'ssh' not in imp and 'ipmi' not in imp]
+
+			# Add ipmi to be the first implementation used
+			if 'ipmi' in self.impl_list:
+				imp_names[:0] = ['ipmi']
+
+			# Add ssh to be the last implementation used
+			if 'ssh' in self.impl_list:
+				imp_names.append('ssh')
+
+		if cmd not in [ 'on', 'off', 'reset', 'status' ]:
+			raise ParamError(self, 'command', 'must be "on", "off", "reset", "status"')
+
 		for host in self.getHostnames(args):
-			for o in self.call('list.host.interface', [ host ]):
-				if o['interface'] == 'ipmi':
-					self.doPower(host, o['ip'], cmd)
+			debug_msgs = []
+
+			# msgs gathers the normal output of the command
+			msgs = []
+			self.beginOutput()
+
+			for imp in imp_names:
+				debug_msgs.append(f'Attempting to set power via {imp}')
+				try:
+					imp_msg = self.runImplementation(imp, [host, cmd])
+					if imp_msg:
+						msgs.append(str(imp_msg))
+					self.mq_publish(host, cmd)
+
+					# Set the imp_success flag since a CommandError
+					# wasn't raised when running the implementation
+					# so it succeeded
+					imp_success = True
 					break
+				except CommandError as imp_error:
+					debug_msgs.append(str(imp_error))
+
+			msg_output = []
+			if msgs:
+				msg_output = '\n'.join(msgs)
+
+			# debug_msgs will have always at least one entry
+			# So it doesn't have to be checked for being empty
+			debug_output = '\n'.join(debug_msgs)
+			if self.debug:
+				self.addOutput(host, msg_output)
+				self.addOutput(host, debug_output)
+			elif msgs:
+				self.addOutput(host, msg_output)
+			self.endOutput(padChar='', trimOwner=True)
 
+			# Raise a CommandError if no implementation succeeded
+			if not imp_success:
+				if self.debug:
+					raise CommandError(self, f'Could not set power cmd {cmd} on host {host}:\n{debug_output}')
+				else:
+					raise CommandError(self, f'Could not set power cmd {cmd} on host {host}')
```

Comment:
If all implementations do not work, it will raise a command error. If one implementation however raises a command error, its output is added to the debug string. Currently set host power outputs nothing if there isn't a valid ipmi interface and if there is just outputs what `ipmitool` says.